Enable Custom Lowering for fabs.v8f16 on AVX

david-xl opened the pull request:
[X86]: Enable custom lowering for fabs.v8f16 on AVX

Currently, custom lowering of fabs.v8f16 requires AVX512FP16, which is too restrictive. For v8f16 fabs lowering, no instructions in AVX512FP16 are needed.  Without the fix, horribly inefficient code is generated without AVX512FP16. Note instcombiner generates calls to intrinsics @llvm.fabs.v8f16 when simplifyping AND <8 x half> operations.

david-xl committed:
Enable Custom Lowering for fabs.v8f16 on AVX

	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

david-xl committed:
Enable Custom Lowering for fabs.v8f16 on AVX

david-xl committed:
Enable Custom Lowering for fabs.v8f16 on AVX

	modified:   llvm/lib/Target/X86/X86ISelLowering.cpp
	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

david-xl committed:
Enable Custom Lowering for fabs.v8f16 on AVX

	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

david-xl committed:
Check softfloat setting for fabs.v8f16 custom lowering

david-xl commented:
This fix improves Eigen pcos<Packet8h> performance by ~65% on Haswell.

RKSimon reviewed (commented):
@david-xl - please can you rebase this?

RKSimon commented:
This looks like the merge has gone wrong?

david-xl commented:
> This looks like the merge has gone wrong?

Right -- I messed it up (new to the workflow). Will fix.

david-xl commented:
> This looks like the merge has gone wrong?

Fixed the bad merge. PTAL.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -2238,6 +2238,9 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
   }
 
+  if (!Subtarget.useSoftFloat() && Subtarget.hasAVX())
+    setOperationAction(ISD::FABS, MVT::v8f16, Custom);
```
Please can you move this inside the main "Subtarget.hasAVX()" section, around line 1425?

RKSimon commented on llvm/test/CodeGen/X86/vec_fabs.ll:
```diff
@@ -137,6 +137,45 @@ define <4 x double> @fabs_v4f64(<4 x double> %p) {
+  %nnv = call <8 x half> @llvm.fabs.v8f16(<8 x half> %v)
+  ret <8 x half> %nnv
+}
+declare <8 x half> @llvm.fabs.v8f16(<8 x half> %p)
```
Please can you add v16f16 / v32f16 test coverage as well?

david-xl committed:
Address review comments

 1. Move the code to the common place
 2. Add test coverage for v16f16 and v32f16 FABS lowering

	modified:   llvm/lib/Target/X86/X86ISelLowering.cpp
	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

	modified:   llvm/lib/Target/X86/X86ISelLowering.cpp
	modified:   llvm/test/CodeGen/X86/vec_fabs.ll

david-xl commented:
Addressed review comments. PTAL.

RKSimon commented on llvm/test/CodeGen/X86/vec_fabs.ll:
```diff
@@ -137,6 +139,86 @@ define <4 x double> @fabs_v4f64(<4 x double> %p) {
+  %v = load <16 x half>, ptr %p, align 32
+  %nnv = call <16 x half> @llvm.fabs.v16f16(<16 x half> %v)
+  ret <16 x half> %nnv
+}
```
Just let the script generate the codegen, even if its really poor for pre-AVX512FP16

david-xl replied:
Ok. I will have a followup patch to update the test.

david-xl replied:
thanks. Looks like you have already updated and cleaned up the test.

RKSimon replied:
Yup, we're still doing a really bad job with v16f16/v32f16 on AVX targets - please can you take a look?

david-xl replied:
Will take a look.

RKSimon reviewed (approved):
LGTM with one minor
